### ISSUE TITLE
Hardware Requirements?

### ISSUE THREAD

**Issue by lrzjason** [lzyhha/VisualCloze#3]
Amazing work. After reading the paper, I have a few questions.
How much is the VRAM requirements for the training? Even resized the image as 384 or 512, concatenating mutiple in context example rows seems require quite a lot VRAM and the paper doesn't focus on this part.
And how much VRAM require for the inference such as two in context examples?
The paper is used channel wise concatenation for in context examples and apply 3D rope. If it is concatenated in height, could it use 2d rope? How different between the 3D rope and 2D rope?
In the paper, it said one step generation or single step generation. But I tried demo and try to se one step inference and it output noise image. Do I misunderstand the meaning of one step in the paper?

**Comment by lrzjason** [comment]
Thanks for the quick response.
It is more clear now.
Could you elaborate more why would it switch from height concatenation to temporal concatenation?
What advantage gained from the change?
It performs something like cross frame attention between in context example and the generation, right?
Thanks

**Comment by lzyhha** [comment]
Spatial concat requires each row of images to have the same width (so that it results in a meaningful grid image). When the aspect ratios of the images differ, this can cause noticeable differences in grid resolution. Temporal concat relaxes this constraint by allowing different aspect ratios per row, resulting in grids with more uniform resolution overall.

During training with 8 A100 GPUs, a batch size of 2 requires 50G memory, when using Fully Sharded Data Parallell that shards model parameters, optimizer states and gradients across DDP ranks. 40GB GPUs can also be used when the batch size is 1. Gradient accumulation can be used to achieve a larger batch size.

For inference, it consumes about 40G memory on an A100 GPU. When using a GPU with 40G memory, the model can also perform inference well through our testing. Using fp8 or offload used in [flux](https://github.com/black-forest-labs/flux/tree/main) can further reduce the memory requirements, and we will try it.

**Comment by lrzjason** [comment]
Thanks for the reply and explanation.

**Comment by libai-lab** [comment]
> Spatial concat requires each row of images to have the same width (so that it results in a meaningful grid image). When the aspect ratios of the images differ, this can cause noticeable differences in grid resolution. Temporal concat relaxes this constraint by allowing different aspect ratios per row, resulting in grids with more uniform resolution overall.
> 
> During training with 8 A100 GPUs, a batch size of 2 requires 50G memory, when using Fully Sharded Data Parallell that shards model parameters, optimizer states and gradients across DDP ranks. 40GB GPUs can also be used when the batch size is 1. Gradient accumulation can be used to achieve a larger batch size.
> 
> For inference, it consumes about 40G memory on an A100 GPU. When using a GPU with 40G memory, the model can also perform inference well through our testing. Using fp8 or offload used in [flux](https://github.com/black-forest-labs/flux/tree/main) can further reduce the memory requirements, and we will try it.

Do you support other models such as sdxl, sd1.5, etc., It would be great if it could take up less GPU and run on consumer grade GPUs as well

**Comment by lzyhha** [comment]
We currently do not support other models.

**Comment by by1e11** [comment]
> Spatial concat requires each row of images to have the same width (so that it results in a meaningful grid image). When the aspect ratios of the images differ, this can cause noticeable differences in grid resolution. Temporal concat relaxes this constraint by allowing different aspect ratios per row, resulting in grids with more uniform resolution overall.
> 
> During training with 8 A100 GPUs, a batch size of 2 requires 50G memory, when using Fully Sharded Data Parallell that shards model parameters, optimizer states and gradients across DDP ranks. 40GB GPUs can also be used when the batch size is 1. Gradient accumulation can be used to achieve a larger batch size.
> 
> For inference, it consumes about 40G memory on an A100 GPU. When using a GPU with 40G memory, the model can also perform inference well through our testing. Using fp8 or offload used in [flux](https://github.com/black-forest-labs/flux/tree/main) can further reduce the memory requirements, and we will try it.

Hi, thanks for the explanation, but we got OOM from time to time on a single 48G GPU when trying to replicate your training script. Are we making any mistakes? 

[2025-04-21 14:40:04] Training arguments: {
  "model_name": "flux-dev-fill-lora",
  "data_path": "configs/data/visualcloze_finetune.yaml",
  "cache_data_on_disk": false,
  "results_dir": "./output/visualcloze_1x1_bs1_mbs1_rank256_lr1e-4_512",
  "max_steps": 5000,
  "global_bsz": 1,
  "micro_bsz": 1,
  "load_t5": true,
  "load_clip": true,
  "global_seed": 20240826,
  "num_workers": 4,
  "log_every": 1,
  "ckpt_every": 100,
  "master_port": 18181,
  "model_parallel_size": 1,
  "data_parallel": "fsdp",
  "checkpointing": true,
  "precision": "bf16",
  "grad_precision": "bf16",
  "lr": 0.0001,
  "auto_resume": true,
  "resume": null,
  "init_from": null,
  "model_path": "/root/data/Models/VisualCloze",
  "grad_clip": 2.0,
  "wd": 0.0,
  "qk_norm": false,
  "caption_dropout_prob": 0.1,
  "snr_type": "lognorm",
  "do_shift": false,
  "task_probs": [
    1.0
  ],
  "lora_rank": 256,
  "grid_resolution": 512,
  "masking_loss": false,
  "full_model": false,
  "training_type": "lora",
  "use_model_ema": false,
  "use_model_watermark": false,
  "use_wandb": false,
  "debug": true
}
You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
[2025-04-21 14:40:10] T5 loaded
[2025-04-21 14:40:15] CLIP loaded
Init model
ckpt_path: /root/data/Models/Flux.1-Fill-dev/flux1-fill-dev.safetensors, lora_path: your_lora_path
Loading checkpoint
apply gradient checkpointing
Trainable params:
[2025-04-21 14:42:32] Model Size: 52.57 GB, Total Parameters: 13.14 B, Trainable Parameters: 1.24 B
[2025-04-21 14:42:32] Creating data
[2025-04-21 14:42:32] Global bsz: 1 Local bsz: 1 Micro bsz: 1

**Comment by lzyhha** [comment]
@by1e11 Hello, is it possible to train if the grid_resolution is set to 384?

**Comment by by1e11** [comment]
> [@by1e11](https://github.com/by1e11) Hello, is it possible to train if the grid_resolution is set to 384? We test the memory consumption using the default setting, which is 384.

Thanks for the reply. Almost solved, implementing a Pipeline parallel version and successfully fit on two GPU's, still testing and training. Will report back when it's ready
